Observable.delay(delayErrors=true) drops onNext signals at random when disposed of concurrently

When a subscription to an `Observable.delay(delayErrors=true)` operator is disposed of concurrently to the subscription's emissions, the `onNext` signals it issues to its downstream can be dropped at random.

See the following repro:

```java
Scheduler scheduler = Schedulers.computation();
int maxValue = 10;
int maxAttempts = 1000;
for (int attempt = 0; attempt < maxAttempts; attempt++) {
    ConcurrentLinkedDeque<Notification<Integer>> sink = new ConcurrentLinkedDeque<>();
    SequentialDisposable firstDisposable = new SequentialDisposable();
    CountDownLatch latch = new CountDownLatch(1);

    Observable<Integer> integers = Observable.range(0, maxValue);
    Observable<Notification<Integer>> first = integers.delay(5, TimeUnit.MICROSECONDS, scheduler, true)
            .materialize()
            .doOnNext(sink::add)
            .doOnSubscribe(firstDisposable::replace);
    Observable<?> second = integers.delay(5, TimeUnit.MICROSECONDS, scheduler)
            .doOnNext($ -> firstDisposable.dispose())
            .doOnTerminate(latch::countDown);

    first.subscribe();
    second.subscribe();
    latch.await();

    boolean foundTermination = false;
    int nextExpectedValue = 0;
    List<String> errors = new ArrayList<>();
    List<Notification<Integer>> notifications = List.copyOf(sink);
    for (int i = 0; i < notifications.size(); i++) {
        Notification<Integer> notification = notifications.get(i);
        if (notification.isOnComplete() || notification.isOnError()) {
            foundTermination = true;
        }
        if (notification.isOnNext()) {
            if (foundTermination) {
                errors.add("Found onNext after termination at index " + i);
            }
            int value = notification.getValue();
            if (value != nextExpectedValue) {
                errors.add((errors.size() + 1) + ". Wrong value at index " + i + " (expected " + nextExpectedValue +
                        ", was " + value + ")");
            }
            nextExpectedValue = value + 1;
        }
    }
    if (!errors.isEmpty()) {
        throw new AssertionError(new StringBuilder()
                .append("Attempt ")
                .append(attempt)
                .append(" failed.\n\t\tError(s):\n\t\t\t")
                .append(String.join("\n\t\t\t", errors))
                .append("\n\t\tNotifications: ")
                .append(notifications)
                .toString());
    }
}
```

Here is an example failure:

```
java.lang.AssertionError: Attempt 11 failed.
            Error(s):
        1. Wrong value at index 2 (expected 2, was 3)
        2. Wrong value at index 3 (expected 4, was 5)
        3. Wrong value at index 6 (expected 8, was 9)
        Notifications: [OnNextNotification[0], OnNextNotification[1], OnNextNotification[3], OnNextNotification[5], OnNextNotification[6], OnNextNotification[7], OnNextNotification[9]]
```

I understand that a subscription that is actively producing signals may continue to propagate those signals to the downstream when disposed of concurrently (i.e. disposal is "eventually consistent"). But I would expect that any signals emitted during or after the race would preserve their original ordering. In other words, once `delay()` has caught up to its newly-disposed state and does _not_ emit its `N`th `onNext` signal, it seems wrong for it to go on to emit its `N+1`th signal.

--- a/src/main/java/io/reactivex/rxjava3/internal/operators/flowable/FlowableDelay.java
+++ b/src/main/java/io/reactivex/rxjava3/internal/operators/flowable/FlowableDelay.java
@@ -111,7 +111,9 @@ public final class FlowableDelay<T> extends AbstractFlowableWithUpstream<T, T> {
 
             @Override
             public void run() {
-                downstream.onNext(t);
+                if (!w.isDisposed()) {
+                    downstream.onNext(t);
+                }
             }
         }
 

--- a/src/main/java/io/reactivex/rxjava3/internal/operators/observable/ObservableDelay.java
+++ b/src/main/java/io/reactivex/rxjava3/internal/operators/observable/ObservableDelay.java
@@ -111,7 +111,9 @@ public final class ObservableDelay<T> extends AbstractObservableWithUpstream<T, 
 
             @Override
             public void run() {
-                downstream.onNext(t);
+                if (!w.isDisposed()) {
+                    downstream.onNext(t);
+                }
             }
         }
 

--- a/src/test/java/io/reactivex/rxjava3/internal/operators/flowable/FlowableDelayTest.java
+++ b/src/test/java/io/reactivex/rxjava3/internal/operators/flowable/FlowableDelayTest.java
@@ -20,6 +20,7 @@ import static org.mockito.Mockito.*;
 import java.util.*;
 import java.util.concurrent.*;
 import java.util.concurrent.atomic.*;
+import java.util.concurrent.locks.LockSupport;
 
 import org.junit.*;
 import org.mockito.InOrder;
@@ -28,6 +29,7 @@ import org.reactivestreams.*;
 import io.reactivex.rxjava3.core.*;
 import io.reactivex.rxjava3.exceptions.TestException;
 import io.reactivex.rxjava3.functions.*;
+import io.reactivex.rxjava3.internal.disposables.SequentialDisposable;
 import io.reactivex.rxjava3.internal.functions.Functions;
 import io.reactivex.rxjava3.processors.PublishProcessor;
 import io.reactivex.rxjava3.schedulers.*;
@@ -1029,5 +1031,39 @@ public class FlowableDelayTest extends RxJavaTest {
         })
         .to(TestHelper.<Integer>testConsumer())
         .assertFailureAndMessage(NullPointerException.class, "The itemDelay returned a null Publisher");
+    }
+
+    @Test
+    public void cancelShouldPreventRandomSubsequentEmissions() {
+        for (int attempt = 1; attempt < 100; attempt ++) {
+
+            SequentialDisposable disposable = new SequentialDisposable();
+            ConcurrentLinkedQueue<Integer> sink = new ConcurrentLinkedQueue<>();
+
+            disposable.replace(
+                Flowable.range(1, 10)
+                .delay(1, TimeUnit.MICROSECONDS, Schedulers.computation(), true)
+                .doOnNext(v -> {
+                    if (v == 1) {
+                        Schedulers.computation().scheduleDirect(disposable::dispose);
+                    }
+                    sink.offer(v);
+                })
+            .subscribe());
+
+            LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(1));
+
+            Integer last = null;
+
+            while (!sink.isEmpty()) {
+                Integer current = sink.poll();
+
+                if (last != null && last + 1 != current) {
+                    fail("Emission hole: " + last + " -> " + current);
+                }
+
+                last = current;
+            }
+        }
     }
 }

--- a/src/test/java/io/reactivex/rxjava3/internal/operators/observable/ObservableDelayTest.java
+++ b/src/test/java/io/reactivex/rxjava3/internal/operators/observable/ObservableDelayTest.java
@@ -20,6 +20,7 @@ import static org.mockito.Mockito.*;
 import java.util.*;
 import java.util.concurrent.*;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.LockSupport;
 
 import org.junit.*;
 import org.mockito.InOrder;
@@ -29,6 +30,7 @@ import io.reactivex.rxjava3.core.Observable;
 import io.reactivex.rxjava3.core.Observer;
 import io.reactivex.rxjava3.exceptions.TestException;
 import io.reactivex.rxjava3.functions.*;
+import io.reactivex.rxjava3.internal.disposables.SequentialDisposable;
 import io.reactivex.rxjava3.internal.functions.Functions;
 import io.reactivex.rxjava3.observers.*;
 import io.reactivex.rxjava3.schedulers.*;
@@ -978,4 +980,37 @@ public class ObservableDelayTest extends RxJavaTest {
         .to(TestHelper.<Integer>testConsumer())
         .assertFailureAndMessage(NullPointerException.class, "The itemDelay returned a null ObservableSource");
     }
-}
+
+    @Test
+    public void cancelShouldPreventRandomSubsequentEmissions() {
+        for (int attempt = 1; attempt < 100; attempt ++) {
+
+            SequentialDisposable disposable = new SequentialDisposable();
+            ConcurrentLinkedQueue<Integer> sink = new ConcurrentLinkedQueue<>();
+
+            disposable.replace(
+                Observable.range(1, 10)
+                .delay(1, TimeUnit.MICROSECONDS, Schedulers.computation(), true)
+                .doOnNext(v -> {
+                    if (v == 1) {
+                        Schedulers.computation().scheduleDirect(disposable::dispose);
+                    }
+                    sink.offer(v);
+                })
+            .subscribe());
+
+            LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(1));
+
+            Integer last = null;
+
+            while (!sink.isEmpty()) {
+                Integer current = sink.poll();
+
+                if (last != null && last + 1 != current) {
+                    fail("Emission hole: " + last + " -> " + current);
+                }
+
+                last = current;
+            }
+        }
+    }}

